File: backend/app/core/sessions.py
import secrets
from datetime import datetime, timedelta, timezone
from sqlalchemy import select, delete
from sqlalchemy.ext.asyncio import AsyncSession
from app.models.auth import UserSession

SESSION_COOKIE_NAME = "session_id"
IDLE_TIMEOUT_HOURS = 2
ABSOLUTE_TIMEOUT_HOURS = 12

async def create_user_session(
    user_id: str,
    db: AsyncSession,
    ip_address: str | None = None,
    user_agent: str | None = None
) -> UserSession:
    """
    Create a new database-backed session token for a user.
    """
    token = secrets.token_urlsafe(32)
    now = datetime.now(timezone.utc)
    expires_at = now + timedelta(hours=IDLE_TIMEOUT_HOURS)

    session = UserSession(
        user_id=user_id,
        session_token=token,
        expires_at=expires_at,
        ip_address=ip_address,
        user_agent=user_agent
    )
    db.add(session)
    await db.commit()
    await db.refresh(session)
    return session

async def validate_user_session(token: str, db: AsyncSession) -> UserSession | None:
    """
    Validate the session token, enforce idle timeout updates, and return the session.
    """
    result = await db.execute(select(UserSession).where(UserSession.session_token == token))
    session = result.scalar_one_or_none()
    if not session:
        return None

    now = datetime.now(timezone.utc)
    # Check if session has expired
    if session.expires_at < now:
        await delete_user_session(token, db)
        return None

    # Check absolute timeout limit (12 hours from created_at)
    if session.created_at + timedelta(hours=ABSOLUTE_TIMEOUT_HOURS) < now:
        await delete_user_session(token, db)
        return None

    # Update idle expiration time
    session.expires_at = now + timedelta(hours=IDLE_TIMEOUT_HOURS)
    session.updated_at = now
    await db.commit()
    return session

async def delete_user_session(token: str, db: AsyncSession) -> None:
    """
    Invalidate session token from database.
    """
    await db.execute(delete(UserSession).where(UserSession.session_token == token))
    await db.commit()

async def cleanup_expired_sessions(db: AsyncSession) -> None:
    """
    Prune all expired sessions from the database.
    """
    now = datetime.utcnow()
    await db.execute(delete(UserSession).where(UserSession.expires_at < now))
    await db.commit()

File: backend/src/auth/sessions.ts
import crypto from 'crypto';
import pg from 'pg';
import { config } from '../config/index.js';
import { pool } from '../database/index.js';

export interface UserSession {
  id: string;
  user_id: string;
  session_token: string;
  expires_at: Date;
  ip_address: string | null;
  user_agent: string | null;
  created_at: Date;
  updated_at: Date;
}

export async function createSession(
  userId: string,
  ipAddress: string | null,
  userAgent: string | null,
  client?: pg.PoolClient
): Promise<UserSession> {
  const db = client || pool;
  const token = crypto.randomBytes(32).toString('base64url');
  const now = new Date();
  const expiresAt = new Date(now.getTime() + config.IDLE_TIMEOUT_MS);

  const query = `
    INSERT INTO user_sessions (id, user_id, session_token, expires_at, ip_address, user_agent)
    VALUES (gen_random_uuid(), $1, $2, $3, $4, $5)
    RETURNING id, user_id, session_token, expires_at, ip_address, user_agent, created_at, updated_at
  `;
  const res = await db.query(query, [userId, token, expiresAt, ipAddress, userAgent]);
  return res.rows[0];
}

export async function validateSession(token: string, client?: pg.PoolClient): Promise<UserSession | null> {
  const db = client || pool;
  const query = `
    SELECT id, user_id, session_token, expires_at, ip_address, user_agent, created_at, updated_at
    FROM user_sessions
    WHERE session_token = $1 AND deleted_at IS NULL
  `;
  const res = await db.query(query, [token]);
  const session = res.rows[0] as UserSession | undefined;

  if (!session) return null;

  const now = new Date();

  // Enforce idle timeout
  if (new Date(session.expires_at).getTime() < now.getTime()) {
    await deleteSession(token, db);
    return null;
  }

  // Enforce absolute timeout (12 hours)
  if (new Date(session.created_at).getTime() + config.ABSOLUTE_TIMEOUT_MS < now.getTime()) {
    await deleteSession(token, db);
    return null;
  }

  // Update idle timeout expiration
  const newExpiresAt = new Date(now.getTime() + config.IDLE_TIMEOUT_MS);
  await db.query(
    'UPDATE user_sessions SET expires_at = $1, updated_at = $2 WHERE session_token = $3',
    [newExpiresAt, now, token]
  );
  session.expires_at = newExpiresAt;

  return session;
}

export async function deleteSession(token: string, client?: pg.PoolClient | pg.Pool): Promise<void> {
  const db = client || pool;
  await db.query('UPDATE user_sessions SET deleted_at = NOW() WHERE session_token = $1', [token]);
}
